src/signTx.c

```diff
@@ -693,45 +786,27 @@ static void signTx_handleInputAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t w
 		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
 	}
 
-	// parsed data
-	struct {
-		uint8_t txHashBuffer[TX_HASH_LENGTH];
-		uint32_t parsedIndex;
-	} input;
-
-	{
-		// parse input
-		TRACE_BUFFER(wireDataBuffer, wireDataSize);
-
-		struct {
-			uint8_t txHash[TX_HASH_LENGTH];
-			uint8_t index[4];
-		}* wireUtxo = (void*) wireDataBuffer;
-
-		VALIDATE(wireDataSize == SIZEOF(*wireUtxo), ERR_INVALID_DATA);
+	BODY_CTX->stageData.input = extractTransactionInput(wireDataBuffer, wireDataSize);
 
-		memmove(input.txHashBuffer, wireUtxo->txHash, SIZEOF(input.txHashBuffer));
-		input.parsedIndex = u4be_read(wireUtxo->index);
-	}
+	security_policy_t policy = policyForSignTxInput(ctx->commonTxData.txSigningMode);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
 
 	{
 		// add to tx
 		TRACE("Adding input to tx hash");
 		txHashBuilder_addInput(
 		        &BODY_CTX->txHashBuilder,
-		        input.txHashBuffer, SIZEOF(input.txHashBuffer),
-		        input.parsedIndex
+		        BODY_CTX->stageData.input.txHashBuffer, SIZEOF(BODY_CTX->stageData.input.txHashBuffer),
+		        BODY_CTX->stageData.input.parsedIndex
 		);
 	}
 
-	security_policy_t policy = policyForSignTxInput();
-	TRACE("Policy: %d", (int) policy);
-	ENSURE_NOT_DENIED(policy);
-
 	{
 		// select UI steps
 		switch (policy) {
 #	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_INPUT_STEP_RESPOND);
```

Comment:
```suggestion
			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_INPUT_STEP_DISPLAY);
```
Please click through the integration test without headless mode to make sure the UI looks as it is supposed to.